### PR TITLE
Refactor to use the windows installer when creating a base environment

### LINKED ISSUE #142: Use the windows installer for installing base python environments
#### Description:

The embeddable python package does not contain tkinter - a necessary library for users that may want to do UI work in python.
Annoyingly tkinter can't be installed directly through pip, so the only other option is to invoke the windows installer to include pip with the base python environment.

Using the windows installer has the side effect of creating registry keys for each major version - meaning that more than one specific version of python cannot be installed at once (e.g 3.10.5 and 3.10.4 cannot be installed at the same time) if using the windows installer. 
As specific versions of python differ little from another within the same major version (usually only bug fixes), we can safely ignore the smaller specific versions and simply use just the most recent version (x.y.z where z is the largest possible number).

This will change the structure of the base python environment - to have each major version of python (if a user installs it), but we will continue to maintain that python 3.10 is the currently supported BHoM base version. The Python_Toolkit folder in PythonEnvironments will contain folders with each version of python currently installed.

Virtual environments will be created from the necessary major version number and ignore the minor version number.

### PR DESCRIPTION
**Pull request by Tom-Kingstone**
### Issues addressed by this PR

Closes #142 

The changes in this PR necessitate changes to the following downstream repositories:
* https://github.com/BHoM/MachineLearning_Toolkit - though this toolkit isn't in alphas/betas so will not be installer checked.
* https://github.com/BHoM/LadybugTools_Toolkit

Refactor of very core methods, so this PR requires extensive testing, with any python based unit tests and the beta testing procedures.

Updated the base python environment logic to use the windows installer instead of embedded packages (which allows tkinter and other similar libraries to be included), and updated the Remove methods respectively to use the installer to remove the environment (as the windows installer modifies the registry to keep track of the version).

Updated relevant enums and dictionaries to only allow the most recent specific version of each minor python version, as the windows installer does not support having more than one specific version downloaded at the same time.

Move the bhom analytics logger from LadybugTools_Toolkit to Python_Toolkit (as this is where it should be to be accessible to all python based toolkits)

### Test files

Use the beta test procedure and any supporting unit tests for this repo and any downstream repositories

### Changelog

* Updated base environment install method to use the windows installer rather than the embeddable package

### Additional comments

When testing, make sure that if you wish to remove an environment, use the methods that remove the environments rather than manually deleting the folder, as removing the folder will not update any registry keys and therefore make it impossible to install that version of python again (this is a limitation of the windows installer). If you accidentally do this, contact me for help to resolve the issue.

### CODE REVIEW

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer

**Comment by Tom-Kingstone** [comment]
@BHoMBot check core
@BHoMBot check compliance

File: Python_Engine/Query/Directory.cs [hunk 1]
```diff
@@ -67,11 +67,12 @@ public static string DirectoryEnvironments()
             return dir;
         }
 
+        [PreviousVersion("8.0", "BH.Engine.Python.Query.DirectoryBaseEnvironment()")]
         [Description("The location where the base Python environment exists.")]
         [Output("The location where the base Python environment exists.")]
-        public static string DirectoryBaseEnvironment()
+        public static string DirectoryBaseEnvironment(PythonVersion version = PythonVersion.v3_10)
```
Comment (Tom-Kingstone) [inline Python_Engine/Query/Directory.cs:73]:
```suggestion
        public static string DirectoryBaseEnvironment(this PythonVersion version)
```

**Comment by Tom-Kingstone** [comment]
@BHoMBot check code-compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check core
@BHoMBot check compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning
@BHoMBot check installer

File: Python_Engine/Compute/BasePythonEnvironment.cs [hunk 2]
```diff
@@ -33,11 +33,14 @@ namespace BH.Engine.Python
 {
     public static partial class Compute
     {
+        [PreviousVersion("8.0", "BH.Engine.Python.Compute.BasePythonEnvironment(System.String, System.String)")]
```
Comment (Tom-Kingstone) [inline Python_Engine/Compute/BasePythonEnvironment.cs:36]:
```suggestion
        [PreviousVersion("8.0", "BH.Engine.Python.Compute.BasePythonEnvironment(System.Boolean, System.Boolean)")]
```
must have been very distracted when I wrote this 😆

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning
@BHoMBot check compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer

**Comment by Tom-Kingstone** [comment]
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation

**Review by CKBoulter — COMMENTED** [review]
Testing procedure for Python_Toolkit is now outdated due to the updates therefore this should be adjusted. Changes to be made should be:

1. Update BasePythonEnvironment to the new component to allow for PythonVersion selection (v3_10).
2. Plug env into _value.
3. Update startJupyterLab script to the correct pathway.

These make the outputs True, however, startJupyterLab no longer launches Jupyter.

**Review by CKBoulter — DISMISSED** [review]
Testing procedure for Python_Toolkit is now outdated due to the updates therefore this should be adjusted. Changes to be made should be:

1. Update BasePythonEnvironment to the new component to allow for PythonVersion selection (v3_10).
2. Plug env into _value.
3. Update startJupyterLab script to the correct pathway.

These make the outputs True, however, startJupyterLab no longer launches Jupyter.

**Comment by Tom-Kingstone** [comment]
@BHoMBot check core
@BHoMBot check compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer
@BHoMBot check versioning

**Comment by Tom-Kingstone** [comment]
@BHoMBot check null-handling
@BHoMBot check serialisation

**Comment by Tom-Kingstone** [comment]
@jamesramsden-bh suggested that the error should also display which base environments are currently installed (if any) as well as clearly stating that the one requested isn't installed.

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer -force

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer

**Comment by Tom-Kingstone** [comment]
@BHoMBot check required

**Comment by Tom-Kingstone** [comment]
@BHoMBot check required

**Comment by Tom-Kingstone** [comment]
@BHoMBot check installer

**Comment by Tom-Kingstone** [comment]
@BHoMBot check required
@BHoMBot check copyright-compliance

**Review by tg359 — APPROVED** [review]
All functionality works as expected. Jupyter still works and registers venvs correctly too 😄 

I have one bit of a feedback though. I understand why we might want to expose different versions to the virtual environments, but for the root environment I think we should hide that option. I tried installing another "BasePythonEnvironment" using a different version and got an error at the very end (trace at the bottom of this message). I think it would be safer to just set the version of the BaseEnvironment behind the scenes and force users to use a specific version to save there being multiple "bases" and some confusion about why this error is happening. This is not a typical install location, nor typical behaviour of a user, so I approve this PR as nothing bad actually happens if you actually install two "Base" environments.

```bash
Obtaining file:///C:/ProgramData/BHoM/Extensions/PythonCode/Python_Toolkit
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... error

[process exited with code 1 (0x00000001)]
You can now close this terminal with Ctrl+D, or press Enter to restart.
Obtaining file:///C:/ProgramData/BHoM/Extensions/PythonCode/Python_Toolkit
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... error
  error: subprocess-exited-with-error

  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      Traceback (most recent call last):
        File "C:\ProgramData\BHoM\Extensions\PythonEnvironments\Python_Toolkit\v3_11\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\ProgramData\BHoM\Extensions\PythonEnvironments\Python_Toolkit\v3_11\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\ProgramData\BHoM\Extensions\PythonEnvironments\Python_Toolkit\v3_11\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 132, in get_requires_for_build_editable
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\tgerrish\AppData\Local\Temp\pip-build-env-91r7q8ug\overlay\Lib\site-packages\setuptools\build_meta.py", line 482, in get_requires_for_build_editable
          return self.get_requires_for_build_wheel(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\tgerrish\AppData\Local\Temp\pip-build-env-91r7q8ug\overlay\Lib\site-packages\setuptools\build_meta.py", line 333, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\tgerrish\AppData\Local\Temp\pip-build-env-91r7q8ug\overlay\Lib\site-packages\setuptools\build_meta.py", line 303, in _get_build_requires
          self.run_setup()
        File "C:\Users\tgerrish\AppData\Local\Temp\pip-build-env-91r7q8ug\overlay\Lib\site-packages\setuptools\build_meta.py", line 521, in run_setup
          super().run_setup(setup_script=setup_script)
        File "C:\Users\tgerrish\AppData\Local\Temp\pip-build-env-91r7q8ug\overlay\Lib\site-packages\setuptools\build_meta.py", line 319, in run_setup
          exec(code, locals())
        File "<string>", line 4, in <module>
      ModuleNotFoundError: No module named 'win32api'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 23.1.2 -> 24.3.1
[notice] To update, run: C:\ProgramData\BHoM\Extensions\PythonEnvironments\Python_Toolkit\v3_11\python.exe -m pip install --upgrade pip

[process exited with code 1 (0x00000001)]
You can now close this terminal with Ctrl+D, or press Enter to restart.
```

**Comment by Tom-Kingstone** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**